편의점 바텀시트 구현

WhiteHyun opened the pull request:
## Screenshots 📸

|UI|
|:-:|
|![RPReplay_Final1708673376](https://github.com/PyeonHaeng/PyeonHaeng-iOS/assets/57972338/57ec1d9d-d174-403e-b59e-62c49d979ad6)|

## 고민, 과정, 근거 💬

### modal로 띄우기 위해 `isPresented`를 어디에 두어야할까

![image](https://github.com/PyeonHaeng/PyeonHaeng-iOS/assets/57972338/cc066051-f313-403b-80f6-3c1ac77b7727)

`Source of truth`를 ViewModel의 state로 한정하려고 했으나, 뷰를 모달로 띄우기 위해선 state를 수정할 권한을 뷰에게 주어야했습니다.
저는 state 변경을 View에서 하지 않고 ViewModel에서만 수정할 수 있도록 제한하고 싶어 결국 새로운 `Source of truth`를 추가했습니다.
다른 좋은 방법이 있다면 언제든 피드백 부탁드리겠습니다.. 🙇

---

- Closed: #47

WhiteHyun committed:
:sparkles: Implement convenience store selection bottom sheet view

WhiteHyun committed:
:sparkles: Implement convenience store selection modal

- Added a bottom sheet view to display the convenience store selection modal.

WhiteHyun committed:
:art: Refactor hardcoded constants to Metrics enum

EnnzoKim commented on PyeonHaeng-iOS/Sources/Scenes/HomeScene/BottomSheet/ConvenienceSelectBottomSheetView.swift:
```diff
@@ -0,0 +1,104 @@
+
+// MARK: - ConvenienceSelectBottomSheetView
+
+struct ConvenienceSelectBottomSheetView<ViewModel>: View where ViewModel: HomeViewModelRepresentable {
+  @EnvironmentObject private var viewModel: ViewModel
+  @Binding private var isPresented: Bool
+
+  init(isPresented: Binding<Bool>) {
+    _isPresented = isPresented
+  }
+
+  var body: some View {
+    VStack(spacing: Metrics.itemSpacing) {
+      Text("편의점 브랜드 선택")
+        .font(.h3)
+        .frame(maxWidth: .infinity, alignment: .leading)
+        .padding(.horizontal, Metrics.horizontalPadding)
+
+      ForEach(ConvenienceStore.allCases, id: \.self) { store in
+        Button {
+          viewModel.trigger(.changeConvenienceStore(store))
+          isPresented = false
+        } label: {
```
P2.
현재 이 방법은 `BottomSheet`와 `DetailSelectionView`간의 의존성이 생기는 것을 우려하여, 혹시 이 방법은 어떠신가요?

```suggestion
struct ConvenienceSelectBottomSheetView<ViewModel>: View where ViewModel: HomeViewModelRepresentable {
  @EnvironmentObject private var viewModel: ViewModel
  @Environment(\.dismiss) var dismiss

  var body: some View {
    VStack(spacing: Metrics.itemSpacing) {
      Text("편의점 브랜드 선택")
        .font(.h3)
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, Metrics.horizontalPadding)

      ForEach(ConvenienceStore.allCases, id: \.self) { store in
        Button {
          viewModel.trigger(.changeConvenienceStore(store))
          dismiss()
        } label: {
```

WhiteHyun replied:
`isPresented`값이 true이면, 나중에 뷰가 업데이트될 때마다 계속 보여질줄 알았는데 아니었네요..! 제가 아직 SwiftUI 동작 방법을 잘 모르나봐요. 하하
얼른 수정하고 공부하러 갑니다~!!

EnnzoKim commented on PyeonHaeng-iOS/Sources/Scenes/HomeScene/HomeView.swift:
```diff
@@ -12,6 +12,7 @@ import SwiftUI
 
 struct HomeView<ViewModel>: View where ViewModel: HomeViewModelRepresentable {
   @StateObject private var viewModel: ViewModel
+  @State private var isPresented: Bool = false
```
P3.
혹시 `isPresented`라는 Source Of Truth를 `DetailSelectionView`에서 정의하면 
다른 문제가 발생할 것을 우려하여 이곳에 정의하신건가요?!

WhiteHyun replied:
최대한 한 곳에서 관리하고 싶었어요. 상위뷰와 하위뷰 각각마다 Source of truth를 갖기에는 관리하기 힘들 거라 판단했거든요.

WhiteHyun replied:
근데 다른 바텀시트뷰를 띄우려면 어쩔 수 없이 다른 Presented 변수를 설정해야하겠네요.. 결국 새로운 Source of truth를 추가해야할 것 같아요. 이 방법이 과연 올바른지는 모르겠으나, 우선 구현해보고 해결방법을 찾아볼게요. 🥹
HomeView 말고 뷰를 띄워주는 주체인 `HomeProductDetailSelectionView`에게 위임하도록 구현할게요!

EnnzoKim reviewed (changes requested):
맞아요.. 상태변수는 어쩔 수 없이 View에 정의하는 것이 가장 깔끔한 것 같아요!
수고하셨습니다! 👍

WhiteHyun committed:
:art: Refactor `isPresented` state to subview

- Moved isPresented state variable to HomeProductDetailSelectionView for better state management.
- Renamed isPresented variable to convenienceStoreModalPresented for improved clarity and readability.

WhiteHyun committed:
:recycle: Refactor modal presentation using `@Environment(\.dismiss)`

- Replaced `@Binding` with `@Environment(\.dismiss)` for modal presentation management.

EnnzoKim reviewed (approved):
수고하셨습니다..! 😀
